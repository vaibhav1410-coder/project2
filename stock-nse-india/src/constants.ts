export const API_RESPONSE_VALIDATION='API Response Validation'
export const IS_TYPE_STRICT=false
