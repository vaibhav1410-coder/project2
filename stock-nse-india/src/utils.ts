import * as Moment from 'moment';
import { extendMoment } from 'moment-range';

const moment = extendMoment(Moment);
/**
 * @private
 */
export const getDateRangeChunks = (startDate: Date, endDate: Date, chunkInDays: number) => {
    const range = moment.range(startDate, endDate)
    const chunks = Array.from(range.by('days', { step: chunkInDays }))
    const dateRanges = []
    for (let i = 0; i < chunks.length; i++) {
        dateRanges.push({
            start: i > 0 ? chunks[i].add(1, 'day').format('DD-MM-YYYY') : chunks[i].format('DD-MM-YYYY'),
            end: chunks[i + 1] ? chunks[i + 1].format('DD-MM-YYYY') : range.end.format('DD-MM-YYYY')
        })
    }
    return dateRanges
}
/**
 * @private
 */
export const sleep = (ms: number) => {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve('')
        }, ms)
    })
}

/**
 * @private
 * @param obj 
 * @returns 
 */
export const getDataSchema = (data: any, isTypeStrict=true): any[] | string => {
    if (typeof data !== 'object')
        return isTypeStrict ? `${typeof data}` : 'any'
    
    if(Array.isArray(data) && typeof data[0] !== 'object'){
        return isTypeStrict ? `${typeof data[0]}[]` : 'any'
    }

    return Object.entries(data).map(([key, value]) => {
        if (Moment.isDate(value))
            return `${key}: ${isTypeStrict ? 'Date': 'any'}`

        if (value === null || typeof value === 'string')
            return `${key}: ${isTypeStrict ? 'string | null': 'any'}`

        if (typeof value !== 'string' && Array.isArray(value)) {
            const typeForEmpty = isTypeStrict ? [] : 'any'
            return {
                [`${key}`]: value.length ? getDataSchema(value[0],isTypeStrict) : typeForEmpty
            }
        }

        if (typeof value === 'object') {
            return {
                [`${key}`]: getDataSchema(value,isTypeStrict)
            }
        }

        return `${key}: ${isTypeStrict ? typeof value : 'any'}`
    })
}
