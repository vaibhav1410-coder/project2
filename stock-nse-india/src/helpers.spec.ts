import {
    getGainersAndLosersByIndex,
    getMostActiveEquities
} from './helpers'
import { getDataSchema } from './utils'
import { API_RESPONSE_VALIDATION, IS_TYPE_STRICT } from './constants'

describe('Helpers', () => {
    test('getGainersAndLosersByIndex', async () => {
        const data = await getGainersAndLosersByIndex("NIFTY 50")
        expect(getDataSchema(data, IS_TYPE_STRICT)).toMatchSnapshot(API_RESPONSE_VALIDATION)
        const isLosers = data.losers.every((equityDetails) => equityDetails.pChange <= 0)
        const isGainers = data.gainers.every((equityDetails) => equityDetails.pChange > 0)
        expect(isLosers).toBeTruthy()
        expect(isGainers).toBeTruthy()
    })

    test('getMostActiveEquities', async () => {
        const data = await getMostActiveEquities("NIFTY 50")
        expect(getDataSchema(data, IS_TYPE_STRICT)).toMatchSnapshot(API_RESPONSE_VALIDATION)
        expect(data.byVolume[0] >= data.byVolume[1]).toBeTruthy()
        expect(data.byValue[0] >= data.byValue[1]).toBeTruthy()
    })
})
