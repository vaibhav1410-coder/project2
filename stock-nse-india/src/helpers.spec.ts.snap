// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers getGainersAndLosersByIndex: API Response Validation 1`] = `
Array [
  Object {
    "gainers": Array [
      "priority: any",
      "symbol: any",
      "identifier: any",
      "series: any",
      "open: any",
      "dayHigh: any",
      "dayLow: any",
      "lastPrice: any",
      "previousClose: any",
      "change: any",
      "pChange: any",
      "totalTradedVolume: any",
      "totalTradedValue: any",
      "lastUpdateTime: any",
      "yearHigh: any",
      "ffmc: any",
      "yearLow: any",
      "nearWKH: any",
      "nearWKL: any",
      "perChange365d: any",
      "date365dAgo: any",
      "chart365dPath: any",
      "date30dAgo: any",
      "perChange30d: any",
      "chart30dPath: any",
      "chartTodayPath: any",
      Object {
        "meta": Array [
          "symbol: any",
          "companyName: any",
          "industry: any",
          Object {
            "activeSeries": "any",
          },
          Object {
            "debtSeries": "any",
          },
          Object {
            "tempSuspendedSeries": "any",
          },
          "isFNOSec: any",
          "isCASec: any",
          "isSLBSec: any",
          "isDebtSec: any",
          "isSuspended: any",
          "isETFSec: any",
          "isDelisted: any",
          "isin: any",
        ],
      },
    ],
  },
  Object {
    "losers": Array [
      "priority: any",
      "symbol: any",
      "identifier: any",
      "series: any",
      "open: any",
      "dayHigh: any",
      "dayLow: any",
      "lastPrice: any",
      "previousClose: any",
      "change: any",
      "pChange: any",
      "totalTradedVolume: any",
      "totalTradedValue: any",
      "lastUpdateTime: any",
      "yearHigh: any",
      "ffmc: any",
      "yearLow: any",
      "nearWKH: any",
      "nearWKL: any",
      "perChange365d: any",
      "date365dAgo: any",
      "chart365dPath: any",
      "date30dAgo: any",
      "perChange30d: any",
      "chart30dPath: any",
      "chartTodayPath: any",
      Object {
        "meta": Array [
          "symbol: any",
          "companyName: any",
          "industry: any",
          Object {
            "activeSeries": "any",
          },
          Object {
            "debtSeries": "any",
          },
          Object {
            "tempSuspendedSeries": "any",
          },
          "isFNOSec: any",
          "isCASec: any",
          "isSLBSec: any",
          "isDebtSec: any",
          "isSuspended: any",
          "isETFSec: any",
          "isDelisted: any",
          "isin: any",
        ],
      },
    ],
  },
]
`;

exports[`Helpers getMostActiveEquities: API Response Validation 1`] = `
Array [
  Object {
    "byVolume": Array [
      "priority: any",
      "symbol: any",
      "identifier: any",
      "open: any",
      "dayHigh: any",
      "dayLow: any",
      "lastPrice: any",
      "previousClose: any",
      "change: any",
      "pChange: any",
      "ffmc: any",
      "yearHigh: any",
      "yearLow: any",
      "totalTradedVolume: any",
      "totalTradedValue: any",
      "lastUpdateTime: any",
      "nearWKH: any",
      "nearWKL: any",
      "perChange365d: any",
      "date365dAgo: any",
      "chart365dPath: any",
      "date30dAgo: any",
      "perChange30d: any",
      "chart30dPath: any",
      "chartTodayPath: any",
    ],
  },
  Object {
    "byValue": Array [
      "priority: any",
      "symbol: any",
      "identifier: any",
      "open: any",
      "dayHigh: any",
      "dayLow: any",
      "lastPrice: any",
      "previousClose: any",
      "change: any",
      "pChange: any",
      "ffmc: any",
      "yearHigh: any",
      "yearLow: any",
      "totalTradedVolume: any",
      "totalTradedValue: any",
      "lastUpdateTime: any",
      "nearWKH: any",
      "nearWKL: any",
      "perChange365d: any",
      "date365dAgo: any",
      "chart365dPath: any",
      "date30dAgo: any",
      "perChange30d: any",
      "chart30dPath: any",
      "chartTodayPath: any",
    ],
  },
]
`;
